### PR TITLE
Add CompactMap functions

### PR DESCRIPTION
**Pull request by MahdiBM**

### CODE REVIEW

**Comment by MahdiBM** [comment]
@adam-fowler Just so you know, an `AsyncSemaphore` test seems flaky:
https://github.com/adam-fowler/async-collections/actions/runs/6940578147/job/18879760918

**Comment by MahdiBM** [comment]
The failure in nightly-main is also the same semaphore problem. The link i shared above though is in Swift 5.7 CI, so we can't say this is a new problem.

**Comment by adam-fowler** [comment]
> @adam-fowler Just so you know, an `AsyncSemaphore` test seems flaky: https://github.com/adam-fowler/async-collections/actions/runs/6940578147/job/18879760918
> 
> I wonder if you'd want to deprecate it and remove it before 1.0.0.

Think I fixed semaphore test

**Comment by MahdiBM** [comment]
@adam-fowler it is kind of a weird situation that Swift itself doesn't somehow support some form of solution for these cases ...

Those were what I used to do with a async map then filter/comapct/flatten, but it's certainly nicer to be able to do the transformation in just one go, specially for users less "good at swift".

It's also worth noting that these dedicated async implementations theoretically should have some better performance than just mapping.

I understand your concern though, about "how far do we want to go into adding all these async function? Where to draw the line?", it's a hard question but I've leant on the side that adding these other 3 main functions is fine.
To be honest i'm thinking it's suboptimal but still fine to add new async functions for each existing std library "main" "closure-based" function, just as someone feels the need and it's somewhat reasonable.

Let me know what you think. I'm not aware of any public WIP for some stuff like this in Swift itself.

**Comment by MahdiBM** [comment]
@adam-fowler how should we proceed? if you prefer not to include a few more convenience functions, I can maintain my own fork since I see value in these and we're using them at work 🙂

**Comment by adam-fowler** [comment]
Sorry @MahdiBM I had kinda forgot about these. I'll have another look this week.

File: Sources/AsyncCollections/CompactMap.swift [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    /// closure for all the elements of the Sequence. This allows all the transform closures
+    /// to run concurrently instead of serially. Returns only when the closure has been run
+    /// on all the elements of the Sequence.
+    /// - Parameters:
+    ///   - priority: Task priority for tasks in TaskGroup
+    ///   - isIncluded: An async closure that takes an element of the
+    ///   sequence as its argument and returns a Boolean value indicating
+    ///   whether the element should be included in the returned array.
+    /// - Returns: An array of the elements that `isIncluded` allowed.
+    public func concurrentCompactMap<T: Sendable>(priority: TaskPriority? = nil, _ isIncluded: @Sendable @escaping (Element) async throws -> T?) async rethrows -> [T] {
```
Comment (adam-fowler) [inline Sources/AsyncCollections/CompactMap.swift:46]: Any reason the closure is called `isIncluded` here but `transform` above. The stdlib uses `transform`.

File: Sources/AsyncCollections/CompactMap.swift [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    /// to run concurrently instead of serially. Returns only when the closure has been run
+    /// on all the elements of the Sequence.
+    /// - Parameters:
+    ///   - maxConcurrentTasks: Maximum number of tasks to running at the same time
+    ///   - priority: Task priority for tasks in TaskGroup
+    ///   - isIncluded: An async closure that takes an element of the
+    ///   sequence as its argument and returns a Boolean value indicating
+    ///   whether the element should be included in the returned array.
+    /// - Returns: An array of the elements that `isIncluded` allowed.
+    public func concurrentCompactMap<T: Sendable>(maxConcurrentTasks: Int, priority: TaskPriority? = nil, _ isIncluded: @Sendable @escaping (Element) async throws -> T?) async rethrows -> [T] {
```
Comment (adam-fowler) [inline Sources/AsyncCollections/CompactMap.swift:91]: ditto `_isIncluded` as above

**Review by adam-fowler — CHANGES_REQUESTED** [review]
Looks good, apart from one minor naming issue

**Comment by MahdiBM** [comment]
@adam-fowler this one should be ready. The other PR i also commented in 🙂.

File: Sources/AsyncCollections/CompactMap.swift (hunk 1 above)
Comment (adam-fowler) [reply Sources/AsyncCollections/CompactMap.swift:46]: I know its a minor thing, but why did you resolve this without comment and similarly below

**Comment by adam-fowler** [comment]
Can you update the CI to use swift:5.10. Because the CI logs are so old I can't see why they failed and also it won't let me re-run the CI.

**Comment by MahdiBM** [comment]
@adam-fowler should be ready 🙂

File: Sources/AsyncCollections/CompactMap.swift (hunk 1 above)
Comment (MahdiBM) [reply Sources/AsyncCollections/CompactMap.swift:46]: ah, didn't think you'd expect a comment since i applied the requested changes 😅

Comment (adam-fowler) [reply Sources/AsyncCollections/CompactMap.swift:46]: I don't see the changes. I'm looking on my phone so maybe I've missed them.

Comment (MahdiBM) [reply Sources/AsyncCollections/CompactMap.swift:46]: Opps sorry, apparently forgot to actually push the changes. Will do in a few mins.

**Review by adam-fowler — APPROVED** [review]

**Pull request merged**